### ISSUE TITLE
[HOLD for payment 2025-01-02] [$250] Search - App redirects to DM when submitting expense in Search after deleting expense

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#52477]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**9.0.61-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** applausetester+odijd0349@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Open DM with any user.
3. Submit an expense to the user.
4. Go to Search.
5. Select the expense via checkbox.
6. Click on the dropdown > Delete expense.
7. Delete the expense.
8. Open FAB while stay on Search page.
9. Click Submit expense under Quick action.
10. Enter amount and submit it.
11. Note that app redirects to DM with the user.
12. Go to Search page.
13. Open FAB > Click Submit expense under Quick action.
14. Enter amount and submit it.
15. Note that app stays in Search page in this case (expected result).

## Expected Result:
In Step 11, app should remain in Search page when submitting expense to user after deleting an expense in Search.

## Actual Result:
In Step 11, app redirects to DM when submitting expense to user after deleting an expense in Search.

This issue also happens with Submit expense option in FAB, and only happens when an expense is deleted in Search and submitting another expense.

## Workaround:
Unknown

## Platforms:

- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/fc5309b1-accb-41d5-9a43-572c0f562379

 </details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021861529064896018849</li>
        <li>Upwork Job ID: 1861529064896018849</li>
        <li>Last Price Increase: 2024-11-26</li>
<li>Automatic offers: </li>
<ul>
<li>DylanDylann | Reviewer | 105296071</li>
<li>FitseTLT | Contributor | 105296073</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @abekkala</details>

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Search - App redirects to DM when submitting expense in Search after deleting expense

### What is the root cause of that problem?
We have a code in ReportActionList that opens the report before scrolling to bottom here
https://github.com/Expensify/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/pages/home/report/ReportActionsList.tsx#L426
so when we notifyNewAction created when creating the expense this navigate is called from the report screen below the search page in the stack
### What changes do you think we should make in order to solve the problem?

We need to check if the current report screen is focused before executing it 
https://github.com/Expensify/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/pages/home/report/ReportActionsList.tsx#L422
```
            if (!hasNewestReportActionRef.current || !isFocused) {

```
but if we don't want the scrollToBottom to be called we can add it here
https://github.com/Expensify/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/pages/home/report/ReportActionsList.tsx#L419
```
            if (!isFromCurrentUser || !isFocused) {

```
### What alternative solutions did you explore? (Optional)

**Comment by FitseTLT** [comment]
[Updated](https://github.com/Expensify/App/issues/52477#issuecomment-2473689938)

**Comment by DylanDylann** [comment]
I am reviewing

**Comment by DylanDylann** [comment]
@FitseTLT 

> We need to give it a time before scrollToBottomForCurrentUserAction is called so that we calculate according to the correct values of lastVisibleActionCreated and lastAction are populated by onyx. This will fix the problem from the root cause.

You mean that we need to wait for BE response before calling scrollToBottomForCurrentUserAction

**Comment by DylanDylann** [comment]
@FitseTLT After checking I don't think your RCA is correct in this case

> We need to give it a time before scrollToBottomForCurrentUserAction is called so that we calculate according to the correct values of lastVisibleActionCreated and lastAction are populated by onyx. This will fix the problem from the root cause.

We should go with @gijoe0295 [proposal](https://github.com/Expensify/App/issues/52477#issuecomment-2473703756)

🎀 👀 🎀 C+ Reviewed

**Comment by FitseTLT** [comment]
>You mean that we need to wait for BE response before calling scrollToBottomForCurrentUserAction

Nope not for the BE but we need to wait the addition of the report action in onyx takes effect on our report action list and lastVisibleAction of the report

>@FitseTLT After checking I don't think your RCA is correct in this case

@DylanDylann  My RCA is correct and the problem is only being visible here because we are being navigated away from search page but the problem also happens for report screen on an empty chat (I have explained why it happens for empty chats in my RCA) and we are unnecessarily navigating to the same report here
https://github.com/FitseTLT/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/pages/home/report/ReportActionsList.tsx#L433
```
                console.log('It is navigating');
                Navigation.navigate(ROUTES.REPORT_WITH_ID.getRoute(report.reportID));
```
Here is the debugging demo and you can easily debug it. Please re-review it we should solve the problem from the root cause Thx


https://github.com/user-attachments/assets/d0f25be7-467b-4e6f-9ffc-34e472b2dffb

**Comment by DylanDylann** [comment]
>  but the problem also happens for report screen on an empty chat 

@FitseTLT Could you please detail this additional bug including steps, actual behavior, and expected behavior?

**Comment by DylanDylann** [comment]
> Nope not for the BE but we need to wait the addition of the report action in onyx takes effect on our report action list and lastVisibleAction of the report

As I understand, lastVisibleActionCreated shouldn't be updated in this case because the user created the request on the search page, the lastVisibleActionCreated only should be updated if the user comes back to the chat screen. So using runAfterInteractions as in your proposal seems like a workaround to me. In the scope of this issue, I think the selected proposal is enough to fix this issue

**Comment by FitseTLT** [comment]
>As I understand, lastVisibleActionCreated shouldn't be updated in this case because the user created the request on the search page, the lastVisibleActionCreated only should be updated if the user comes back to the chat screen. 

@DylanDylann You are mixing up lastVisibleActionCreated with lastVisitTime, lastVisibleActionCreated is updated whenever a new visible action is created with `created` timestamp of the action so it should be updated whether it is created from chat screen or not. 
Let me explain the purpose of the navigate here
https://github.com/Expensify/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/pages/home/report/ReportActionsList.tsx#L429-L433
When we are linked to an old reportAction (comment linking) in a chat with many report actions and the most recent actions haven't been fetched/available and scrollToBottom is called we will check if we have the newest report action and if we don't  we will navigate to the chat (without the report action link) to fetch the most recent actions.
So now if you think `runAfterInteraction` is a workaround we can apply my alternative [solution](https://github.com/Expensify/App/issues/52477#issuecomment-2473689938) of only navigating when `isFocused` is true which will solve the problem from the root cause that will solve similar problems too because we only want to navigate (and even to scrollToBottom) to (fetch latest actions) when the screen is focused. Just like we are early returning when the action is not from the current user here
https://github.com/Expensify/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/pages/home/report/ReportActionsList.tsx#L426-L427
that will ensure similar problems will not occur for other cases like [sendInvoice](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L3746) [trackExpense](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L3940) [splitBill](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L4506) [here](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L4573) [createDistanceRequest](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L5320).

**Comment by DylanDylann** [comment]
@FitseTLT Thanks for your explanation

> @DylanDylann You are mixing up lastVisibleActionCreated with lastVisitTime, lastVisibleActionCreated is updated whenever a new visible action is created with created timestamp of the action so it should be updated whether it is created from chat screen or not.

**Comment by DylanDylann** [comment]
Waiting for @danieldoglas to give a final assignment

**Comment by FitseTLT** [comment]
@danieldoglas Please check out my comment [here](https://github.com/Expensify/App/issues/52477#issuecomment-2510215976) I think having `isFocused` check before navigating is better as it will solve the problem for all cases [sendInvoice](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L3746) [trackExpense](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L3940) [splitBill](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L4506) [here](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L4573) [createDistanceRequest](https://github.com/FitseTLT/App/blob/4d9be4eb2537b49fa6cf35135770cf3c6d62a1cd/src/libs/actions/IOU.ts#L5320) as we only need the navigation when the report screen is on focus as I have explained in the comment. Thx 👍

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by danieldoglas** [comment]
@DylanDylann can you confirm you can still reproduce the issue?

**Comment by DylanDylann** [comment]
I can't reproduce anymore. @FitseTLT Could you reproduce this issue?

**Comment by FitseTLT** [comment]
> I can't reproduce anymore. @FitseTLT Could you reproduce this issue?

Yes I do. @DylanDylann You should make sure the chat is with a user you have never chatted before (as I explained in my RCA) 👍 


https://github.com/user-attachments/assets/741f9fa5-9fac-4ec7-8269-2358cfef5455

**Comment by DylanDylann** [comment]
@FitseTLT In your video, you don't use quick action, right?

**Comment by FitseTLT** [comment]
> @FitseTLT In your video, you don't use quick action, right?

No no @DylanDylann the bug has nothing to do with Quick Action here is another repro with it

https://github.com/user-attachments/assets/ec1f66e8-9444-42cc-bdae-290e5198decc

**Comment by muttmuure** [comment]
Not a quality bug

**Comment by danieldoglas** [comment]
Wait, what do you mean it has nothing to do with Quick Action?

The reproduction steps clearly say:

```
- Click Submit expense under Quick action.
```

**Comment by FitseTLT** [comment]
> Wait, what do you mean it has nothing to do with Quick Action?
> 
> The reproduction steps clearly say:
> 
> ```
> - Click Submit expense under Quick action.
> ```

Yeah it has no relation with the RCA of the bug so it can be reproduced with QAB as in [here](https://github.com/Expensify/App/issues/52477#issuecomment-2528446522) or not as in [here](https://github.com/Expensify/App/issues/52477#issuecomment-2528415392)

**Comment by DylanDylann** [comment]
@danieldoglas I just tried again and verified that we can reproduce it 


1. Open DM with a new user who doesn't exist on Expensify 
2. Submit an expense to the user
3. Go to Search.
4. Select the expense (created in step 2) via checkbox.
5. Click on the dropdown > Delete expense.
6. Open FAB while staying on the Search page.
7. Click Submit expense under Quick action.
8. Enter the amount and submit it.
9. See that app redirects to DM with the user.


https://github.com/user-attachments/assets/388578d2-fe40-408b-b8d5-03b29c384612

**Comment by danieldoglas** [comment]
After re-reading the discussion, I think that the solution from @FitseTLT is a better approach that might prevent bugs on future navigations as well. Assigning them.

**Comment by DylanDylann** [comment]
@FitseTLT Please prioritize creating PR before posting new proposal

https://github.com/Expensify/App/blob/c1454f772b8b780e015c01c1bffc4669d5383a78/contributingGuides/CONTRIBUTING.md?plain=1#L143

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by Christinadobrzyn** [comment]
Hi @FitseTLT @DylanDylann can you tell if this blocker is from this PR https://github.com/Expensify/App/issues/54505

**Comment by FitseTLT** [comment]
> Hi @FitseTLT @DylanDylann can you tell if this [blocker](https://github.com/Expensify/App/issues/54505) is from this PR

No @Christinadobrzyn It can not be related to this 👍

**Comment by abekkala** [comment]
@DylanDylann can you complete the checklist above?

**Comment by DylanDylann** [comment]
# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by abekkala** [comment]
@FitseTLT and @DylanDylann payments sent and contracts ended - thank you! 🎉
